kernel/src/actions/mod.rs

```diff
@@ -192,8 +190,16 @@ impl Protocol {
                 )
             );
         }
-        let reader_features = reader_features.map(|f| f.into_iter().map(Into::into).collect());
-        let writer_features = writer_features.map(|f| f.into_iter().map(Into::into).collect());
+        let reader_features: Option<Vec<ReaderFeatures>> = reader_features.and_then(|f| {
+            f.into_iter()
+                .map(|f| ReaderFeatures::from_str(&f.into()).ok())
+                .collect::<Option<Vec<ReaderFeatures>>>()
+        });
```

Comment:
could collect as a result? and prob can infer type
```suggestion
        let reader_features = reader_features.and_then(|features| {
            features.into_iter()
                .map(|f| ReaderFeatures::from_str(&f.into()))
                .collect::<Result<Vec<_>, _>>()
                .ok()
        });
```

Reply:
But we need an option here for the struct, not a result or am I missing something?